Userland/Applications/FileManager/PropertiesWindowHashesTab.cpp

```diff
@@ -0,0 +1,202 @@
+        m_sha1_widget->background_calculate_set_and_update(this->make_weak_ptr<HashTab>());
+        m_sha256_widget->background_calculate_set_and_update(this->make_weak_ptr<HashTab>());
+        m_sha512_widget->background_calculate_set_and_update(this->make_weak_ptr<HashTab>());
+    }
+    GUI::Clipboard::the().on_change = [this](String const& mime_type) -> void {
+        if (mime_type.starts_with("text/")) {
+            this->update_clipboard();
+            this->m_md5_widget->update_status(this->make_weak_ptr<HashTab>());
+            this->m_sha1_widget->update_status(this->make_weak_ptr<HashTab>());
+            this->m_sha256_widget->update_status(this->make_weak_ptr<HashTab>());
+            this->m_sha512_widget->update_status(this->make_weak_ptr<HashTab>());
+        }
+    };
+}
+
+Optional<String> HashTab::hash_text_for_file(Crypto::Hash::HashKind const hash_kind) const
+{
+    auto file_to_hash = Core::File::construct(m_path);
+    if (!file_to_hash->open(Core::OpenMode::ReadOnly))
+        return {};
+
+    Crypto::Hash::Manager hash;
+    hash.initialize(hash_kind);
+    String hash_out_string;
+
+    while (!file_to_hash->eof() && !file_to_hash->has_error())
+        hash.update(file_to_hash->read(PAGE_SIZE));
+    auto digest = hash.digest();
+    auto digest_data = digest.immutable_data();
+    StringBuilder hash_string_builder;
+    for (size_t i = 0; i < hash.digest_size(); i++)
+        hash_string_builder.appendff("{:02x}", digest_data[i]);
+    hash_out_string = hash_string_builder.build();
+    file_to_hash->close();
+
+    return hash_out_string;
+}
+
+void HashTab::update_clipboard()
+{
+    auto const& clipboard_data_and_type = GUI::Clipboard::the().data_and_type();
+    if (clipboard_data_and_type.mime_type.starts_with("text/"))
+        m_clipboard = StringView(clipboard_data_and_type.data).to_string();
+    else
+        m_clipboard = {};
+}
+
+HashTab::HashWidget::HashWidget(WeakPtr<HashTab> containing_tab, String const label, Crypto::Hash::HashKind const hash_kind)
+    : m_hash_kind(hash_kind)
+{
+    VERIFY(!containing_tab.is_null());
+    load_from_gml(properties_window_hashes_hash_widget_gml);
+
+    m_hash_name_label = find_descendant_of_type_named<GUI::Label>("hash_kind_label");
+    VERIFY(!m_hash_name_label.is_null());
+    m_hash_name_label->set_text(label);
+
+    m_status_icon = find_descendant_of_type_named<GUI::ImageWidget>("status_icon");
+    VERIFY(!m_status_icon.is_null());
+    m_status_icon->set_bitmap(RefPtr(containing_tab->m_icon_hash_load));
+
+    m_hash_status_label = find_descendant_of_type_named<GUI::Label>("hash_status_label");
+    VERIFY(!m_hash_status_label.is_null());
+
+    m_calulate_button = find_descendant_of_type_named<GUI::Button>("calculate_button");
+    VERIFY(!m_calulate_button.is_null());
+    m_calulate_button->set_icon(RefPtr(containing_tab->m_icon_calculate));
+    m_calulate_button->on_click = [this, containing_tab](unsigned) -> void {
```

Comment:
There's no official rule about this, but `auto` is prefered in this case.
```suggestion
    m_calulate_button->on_click = [this, containing_tab](auto) -> void {
```